### PR TITLE
[🗄️ cache] Make ApolloStore.publish() suspend

### PR DESCRIPTION
**Pull request by martinbonnin**
Made `publish` suspend so that the cache writers wait for the readers and no cache key event is lost. The buffer of size 64 is still there to smooth things out.

Incubating cache gets removed `publish` parameter.  
Stable cache gets overloads `writeFoo()` functions. The ones with a `publish` parameter should probably be deprecated but I'm not 100% sure yet.

Next steps: 
* benchmark thread hops vs no thread hops
* make `watch()` return a `FusibleFlow` so that the user has control of the buffering there.
* maybe deprecated `ApolloStore.writeFoo(publish)`

See #3920

### CODE REVIEW

File: tests/integration-tests/src/commonTest/kotlin/test/WatcherTest.kt [hunk 1]
```diff
@@ -330,14 +331,17 @@ class WatcherTest {
...
+    val job = launch {
       apolloClient.query(query)
           .watch(null)
           .collect {
             channel.send(it.data)
           }
     }
 
+    // Because subscribe is called from a background thread, give some time to be effective
+    delay(500)
```
Comment (martinbonnin) [inline tests/integration-tests/src/commonTest/kotlin/test/WatcherTest.kt:343]: That's an interesting one. If we want a guarantee that `.watch()` subscribes synchronously, we probably need to go back to "interceptors do their own dispatcher switch". I want to measure how costly the thread switch is.

File: libraries/apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ApolloStore.kt [hunk 2]
```diff
@@ -75,14 +75,30 @@ interface ApolloStore {
...
       operation: Operation<D>,
       operationData: D,
       customScalarAdapters: CustomScalarAdapters = CustomScalarAdapters.Empty,
       cacheHeaders: CacheHeaders = CacheHeaders.NONE,
       publish: Boolean = true,
   ): Set<String>
 
+  /**
+   * Write an operation data to the store and optionally publish changes of [Record] which have changed,
+   * that will notify any watcher that depends on these [Record] to re-fetch.
```
Comment (BoD) [inline libraries/apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ApolloStore.kt:88]:
```suggestion
   * Write an operation data to the store.
```

Or maybe do we want to highlight it a bit more? 

```suggestion
   * Write an operation data to the store. Call [publish] with the returned keys to notify any watcher that depends on these [Record] to re-fetch.
```

(same for `writeFragment`)

**Review by BoD — APPROVED** [review]
⚙️

File: libraries/apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ApolloStore.kt (hunk 2 above)
Comment (martinbonnin) [reply libraries/apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ApolloStore.kt]: Good call out. I added a `@see publish` tag so hopefully it should be enough.

**Pull request merged**